src/util/treadmill.rs

```diff
@@ -6,111 +6,192 @@ use crate::util::ObjectReference;
 
 use super::object_enum::ObjectEnumerator;
 
+/// A data structure for recording objects in the LOS.
+///
+/// It is divided into the nursery and the mature space, and each of them is further divided into
+/// the from-space and the to-space.
+///
+/// All operations are protected by a single mutex [`TreadMill::sync`].
 pub struct TreadMill {
-    from_space: Mutex<HashSet<ObjectReference>>,
-    to_space: Mutex<HashSet<ObjectReference>>,
-    collect_nursery: Mutex<HashSet<ObjectReference>>,
-    alloc_nursery: Mutex<HashSet<ObjectReference>>,
+    sync: Mutex<TreadMillSync>,
+}
+
+/// The synchronized part of [`TreadMill`]
+#[derive(Default)]
+struct TreadMillSync {
+    nursery: SpacePair,
```

Comment:
We normally think `nursery` is the from space and `mature` is the to space. This code now uses two 'logical spaces' (from space and to space) for both nursery and mature -- that is 4 logical spaces. I don't know how that makes sense.

Besides, `collect_nursery` is supposed to hold objects that are allocated during a GC (e.g. an over-sized object is copied to LOS from GenImmix nursery). However, we don't do that any more. Large objects are allocated directly into LOS, and there is no copy from other spaces to LOS. I think you can assert if that is true. If it is true, then the code becomes `nursery + from + to` which is similar to `GenCopy` -- that makes much more sense.

Reply:
It turns out that the original 'treadmill' collector does have 4 spaces: https://cofault.com/treadmill.html. We can move to something closer to that. It is still fine to use 4 sets, instead of 1 doubly linked list.

Reply:
> BTW I don't think we have to do any major refactoring to `TreadMill` to support object enumeration. But if you do plan for any major refactoring (such as `[nursery, mature] x [from, to]`), I don't think it is a good idea to move further away from the original 'treadmill' concept.

I don't plan to do a major refactoring for now.  I just needed to make the mutex more coarse-grained which requires some structural change.  I ended up using just `nursery + from + to` which should be the simplest we can have to support our current LOS usage.

Reply:
Oh wait.  There is one catch.  If we enumerate objects during mutator time, we should enumerate both `nursery` and `to_space` because the nursery can contain newly allocated objects which should be returned, too.  But the "nursery from-space" should be empty.  Because it is only an assertion and it has undefined behavior if the mutator attempts to enumerate objects during GC, I'll reduce the precision of the assertion for now.

Reply:
Yes I remember this discussion on how the treadmill we've implemented is very different from the actual one: https://github.com/mmtk/mmtk-core/issues/517